KAFKA-9156: fix LazyTimeIndex & LazyOffsetIndex concurrency

alexandrfox opened the pull request:
Fix of https://issues.apache.org/jira/browse/KAFKA-9156

Race condition in concurrent  `get` method invocation of lazy indexes might lead to multiple `OffsetIndex`/`TimeIndex` objects being concurrently created. When that happens position of `MappedByteBuffer` in `AbstractIndex` is advanced to the last entry which in turn leads to a critical `BufferOverflowException` being thrown whenever leader or replica tries to append a record to the segment.

Moreover, `file_=` setter is seemingly also vulnerable to the race, since multiple threads can attempt to set a new file reference as well as create new Time/OffsetIndex objects at the same time. This might lead to the discrepant File references being held by e.g. LazyTimeIndex and its TimeIndex counterpart.

This patch attempts to fix the issue by making sure that index objects are atomically constructed when loaded lazily via `get` method. Additionally, `file` reference modifications are also made atomic and thread safe.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
KAFKA-9156: fix lazy indexes concurrency

See https://issues.apache.org/jira/browse/KAFKA-9156

lushilin commented on core/src/main/scala/kafka/log/TimeIndex.scala:
```diff
@@ -240,24 +241,34 @@ object TimeIndex extends Logging {
-      timeIndex.get.file = f
-    else
-      _file = f
+    inLock(lock) {
```
i think this lock is too big.this lock can move to timeIndex is not define, and can use double check

alexandrfox replied:
considering how infrequently this is called I am not sure that additional double-check locking complication is needed here, but I am open to modify it in case others agree as well

lushilin commented on core/src/main/scala/kafka/log/TimeIndex.scala:
```diff
@@ -240,24 +241,34 @@ object TimeIndex extends Logging {
-      timeIndex.get.file
-    else
-      _file
+    inLock(lock) {
```
i think this lock is useless because this is a get file method.if modify the file, also can get correct file at next time.

alexandrfox replied:
I believe getter is also susceptible for race. Consider the following situation:
 - Thread 1 enters "else" codepath of the `file` getter and is ready to return the reference of `_file`
 - At this very moment Thread 2 performs two operations consecutively: calls `get` and immediately `file_=` setter replacing the current `file` reference within `TimeIndex` (but leaving `LazyTimeIndex`'s `_file` outdated
 - Thread 1 continues its operation and returns a stale reference to the File object which is no longer correct

ijuma replied:
I think a better way to handle this is to replace `Option[TimeIndex]` with two case classes that inherit from a common sealed trait. Then we don't have this issue where we have to keep 2 separate fields in sync.

alexandrfox replied:
@ijuma agreed, please check out the latest commit

ijuma commented on core/src/main/scala/kafka/log/OffsetIndex.scala:
```diff
@@ -218,24 +219,34 @@ object OffsetIndex extends Logging {
+        }
+      }
+    }
     offsetIndex.get
```
The usual pattern for something like this is to read to a local var first. And avoid another volatile read for the common case.

alexandrfox replied:
@ijuma good catch, fixed

Commit:
KAFKA-9156: improved lazy indexes hierarchy

Moved common logic of LazyOffsetIndex and LazyTimeIndex into a common
abstract class. Moved file reference modification into a sealed trait to
simplify "file" getters and setters.

alexandrfox commented on core/src/main/scala/kafka/log/AbstractLazyIndex.scala:
```diff
@@ -0,0 +1,85 @@
+  def file: File = lazyIndex.file
+
+  def file_=(f: File): Unit = {
+    inLock(lock) {
```
unfortunately we cannot simply move this setter logic into the trait/case classes as well, since then we'd need to declare `indexFile` as `@volatile var indexFile` in `UninitializedLazyIndex` case class. This won't work for scala 2.11 due to https://github.com/scala/bug/issues/8873

ijuma replied:
We don't need them to be case classes though. A plain old class will do. Also, we should move the sealed trait and subclasses to a companion object since they don't seem to need access to the outer class.

alexandrfox replied:
agreed, moved into companion object

ijuma commented on core/src/main/scala/kafka/log/AbstractLazyIndex.scala:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  def get: T = {
+    var localLazyIndex = lazyIndex
```
With the match based approach, we don't actually need this outer local `var` since the pattern match introduces a local val.

alexandrfox replied:
good to know!

ijuma commented on core/src/main/scala/kafka/log/AbstractLazyIndex.scala:
```diff
@@ -0,0 +1,85 @@
+    }
+  }
+
+  def initializeIndex(indexFile: File): T
```
Should this be `protected`?

ijuma replied:
Also, should this be `createIndex` or `newIndex`? `initializeIndex` sounds like it's taking an index and initializing it.

alexandrfox replied:
++ for protected. I called it `createIndex` initially but felt a bit weird since it doesn't actually create the index file, rather just initialising a wrapper around it.

ijuma replied:
That's a fair point. We should add some scaladoc. Maybe `loadIndex`?

alexandrfox replied:
yeah, I like `loadIndex` as well, — that makes more sense since it complements the "lazy" part, will add the javadoc

ijuma commented:
Thanks for the update, looking better. I left a couple more comments.

ijuma commented on core/src/main/scala/kafka/log/OffsetIndex.scala:
```diff
@@ -216,26 +216,11 @@ object OffsetIndex extends Logging {
-    else
-      _file
-  }
+class LazyOffsetIndex(private val _file: File, baseOffset: Long, maxIndexSize: Int = -1, writable: Boolean = true)
```
We can remove `private val` from `_file` here and in `LazyTimeIndex`. That's the default for classes.

Commit:
KAFKA-9156: AbstractLazyIndex structure improvements

ijuma commented on core/src/main/scala/kafka/log/AbstractLazyIndex.scala:
```diff
@@ -0,0 +1,85 @@
+
+  def file: File = lazyIndex.file
+
+  def file_=(f: File): Unit = {
```
We should probably name this differently since it's doing a lot more work than just assigning the new file.

alexandrfox replied:
this is now simplified since file modification logic was moved into the trait subclasses (I dropped `case class` so now it's possible to use `@volatile` again in scala 2.11), — I believe name still makes sense, how do you think?

alexandrfox commented on core/src/main/scala/kafka/log/AbstractLazyIndex.scala:
```diff
@@ -0,0 +1,85 @@
+ * This defers the index initialization to the time it gets accessed
+ * so the cost of the heavy memory mapped operation gets amortized over time.
+ */
+abstract class AbstractLazyIndex[T <: AbstractIndex[_, _]](private val _file: File) {
```
I am wondering if we should keep this in its own file, or move-in with `AbstractIndex.scala` since both `LazyTimeIndex` and `LazyOffsetIndex` are kept together with `TimeIndex` and `OffsetIndex` definitions. Alternatively we can move them out into their own files. WDYT @ijuma?

ijuma replied:
Makes sense to keep it in the same file as the `Lazy*Index` implementations.

ijuma replied:
Oh, I see. The current structure is a bit weird. I think I'd keep all the Lazy implementations together. If you don't mind, I'll push a commit to your branch and we can keep/remove/adjust after.

alexandrfox replied:
sounds good!

ijuma committed:
Simplify

ijuma commented:
For some reason, I cannot push to your branch, so I submitted a PR: https://github.com/bookingcom/kafka/pull/1/files

Commit:
Merge pull request #1 from ijuma/fix-lazy-indexes-concurrency

Simplify and harden

Commit:
add missing type annotations in AbstractIndex

alexandrfox commented:
> For some reason, I cannot push to your branch, so I submitted a PR: https://github.com/bookingcom/kafka/pull/1/files

Probably due to org. settings. LGTM, so I merged it, thank you! Also, added a few missing type annotations to `_maxEntries` and `_entries` of `AbsrtactIndex`

ijuma commented on core/src/main/scala/kafka/log/LazyIndex.scala:
```diff
@@ -0,0 +1,85 @@
+  *
+  * @param loadIndex A function that takes a `File` pointing to an index and returns a loaded `AbstractIndex` instance.
+  */
+class LazyIndex[T <: AbstractIndex] private (@volatile private var indexWrapper: IndexWrapper, loadIndex: File => T) {
```
We should mention that the methods in this class are thread-safe, but to check the underlying `AbstractIndex` subclass documentation for information on their thread safety.

ijuma reviewed (approved):
LGTM, just a minor documentation suggestion below. It would be good to get a second opinion from @junrao or @hachikuji before we merge this to 2.3, 2.4 and trunk.

ijuma commented:
@alexandrfox In the PR description, it's worth clarifying that mutation operations are done under a lock in the caller classes. I think the gap is that `get` can be called during `Log.read` which does not acquire any lock.

alexandrfox commented:
> @alexandrfox In the PR description, it's worth clarifying that mutation operations are done under a lock in the caller classes. I think the gap is that `get` can be called during `Log.read` which does not acquire any lock.

Makes sense, I think LogCleaner also may call `get` while performing `doClean` (which can naturally happen at the same time as `read`).

Commit:
add thread safety infromation

junrao commented on core/src/main/scala/kafka/log/LazyIndex.scala:
```diff
@@ -0,0 +1,90 @@
+    def file_=(f: File)
+  }
+
+  private class IndexFile(@volatile var file: File) extends IndexWrapper
```
Does file need to be var? It seems that it's never updated.

alexandrfox replied:
hey @junrao, yes, this reference is updated when `LazyIndex.file_=` setter is invoked before `LazyIndex.get`

junrao replied:
@alexandrfox : Thanks. That makes sense. So, we can keep the code as it is.

junrao reviewed (approved):
@alexandrfox : Thanks for the PR. LGTM. Just a minor comment below.

hachikuji reviewed (approved):
LGTM. Nice fix and thanks for the additional cleanup!

alexandrfox commented:
@ijuma looks like we have approvals, thanks to @junrao and @hachikuji. And thank you for taking your time guiding me through my first PR!

Also, thanks to @lushilin for pin-pointing the problem!

ijuma commented:
One job passed, two jobs failed with known flaky tests:

> kafka.admin.ReassignPartitionsClusterTest.shouldListMovingPartitionsThroughApi
> kafka.api.ConsumerBounceTest.testClose
> kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup

ijuma commented:
Merged to trunk, 2.4 and 2.3.
